Add cloud-agnostic CloudClient interface

mbarnes opened the pull request:
This is the completion of @lisa's https://github.com/lisa/cloud-ingress-operator/tree/osd-3566 branch.

This was a collective Team Rocket effort, so apologies for the messy commit history.

All controllers have been updated to use the new `CloudClient` interface.

The utility functions proved hardest to wrangle.  I've introduced a new `pkg/utils` subpackage which conceptually sits "below" the  `pkg/controller` and `pkg/cloudclient` layers so as to avoid import loops.  This is where `GetInfrastructureObject` and most of the utilities that rely on it ended up.

lisa commented:
/hold

(for review and eventual squashing)

lisa commented:
/cc @lisa

lisa reviewed (changes requested):
What is the status of testing in-cluster here?

Love to see this getting to PR state.

Can we consolidate commits as per Slack convo?

mbarnes commented on pkg/cloudclient/aws/private.go:
```diff
@@ -0,0 +1,831 @@
+	return nil
+}
+
+// getAWSDecodedProviderSpec casts the spec.providerSpec of a kubernetes machine
```
:heavy_check_mark:

mbarnes commented on pkg/cloudclient/aws/private.go:
```diff
@@ -0,0 +1,831 @@
+	return infra.Status.PlatformStatus.AWS.Region, nil
+}
+
+// getMasterMachines returns a machineList object whose .Items can be iterated
+// over to perform actions on/with information from each master machine object
+func getMasterMachines(kclient client.Client) (*machineapi.MachineList, error) {
+	machineList := &machineapi.MachineList{}
+	listOptions := []client.ListOption{
+		client.InNamespace("openshift-machine-api"),
+		client.MatchingLabels{masterMachineLabel: "master"},
+	}
+	err := kclient.List(context.TODO(), machineList, listOptions...)
+	if err != nil {
+		return nil, err
+	}
+	return machineList, nil
+}
```
I think it got dragged in with `GetMasterNodeSubnets`, but I'll split it back out.

mbarnes replied:
:heavy_check_mark:

mbarnes commented on pkg/cloudclient/aws/private.go:
```diff
@@ -0,0 +1,831 @@
+	return nil
+}
+
+// AddAWSLBToMasterMachines adds a Load Balancer (with name elbName) to the
+// spec.providerSpec.value.loadBalancers list for each of the master machine
+// objects in a cluster
+func AddAWSLBToMasterMachines(kclient client.Client, elbName string, masterNodes *machineapi.MachineList) error {
```
:heavy_check_mark:

mbarnes commented on pkg/cloudclient/aws/private_test.go:
```diff
@@ -0,0 +1,579 @@
+}
+
+func TestCreateNetworkLoadBalancer(t *testing.T) {
+	// TODO: Fill this in
```
Looks like that comment's been there awhile.  I see a test case so I assume it's completed.

mbarnes replied:
:heavy_check_mark:

mbarnes commented:
I've cleaned up the commit history so it shows a more concise and straight-forward progression, while retaining proper authorship.  Also addressed all review comments so far.  Force-pushing the cleaner commit list.

> What is the status of testing in-cluster here?
Everything seemed to be working before I started messing with commits.  I'll re-verify.

mbarnes commented on pkg/controller/publishingstrategy/publishingstrategy_controller.go:
```diff
@@ -182,193 +179,32 @@ func (r *ReconcilePublishingStrategy) Reconcile(request reconcile.Request) (reco
 			return reconcile.Result{}, err
 		}
-		log.Info(fmt.Sprintf("%s successful", comment))
+		log.Info("Update api.<clusterName> alias to internal NLB successful")
```
:heavy_check_mark:

mbarnes commented on pkg/controller/publishingstrategy/publishingstrategy_controller.go:
```diff
@@ -182,193 +179,32 @@ func (r *ReconcilePublishingStrategy) Reconcile(request reconcile.Request) (reco
-			log.Error(err, "Error adding new LB to master machines' providerSpecs")
-			return reconcile.Result{}, err
-		}
+		log.Info("Update api.<clusterName> alias to internal NLB successful")
```
:heavy_check_mark:

lisa commented:
It is with great pleasure that I get to be the one to do this:

/lgtm

:shipit:   🎉 👍 

Great work to everyone involved with this port.


@mbarnes Please `/unhold` when ready to merge.
